e2e cases failed quite often after testing TestFlowAggregator

**Describe the bug**
Many e2e tests failed after testing TestFlowAggregator regardless of where it ran:
Jenkins tests:
https://jenkins.antrea-ci.rocks/job/antrea-e2e-for-pull-request/2180/console
https://jenkins.antrea-ci.rocks/job/antrea-e2e-for-pull-request/2181/console
Kind tests:
https://pipelines.actions.githubusercontent.com/rI6se88BdHGIt3c6pWSIOE7hniT9sK2Yykq02f3adUskIOWLhd/_apis/pipelines/1/runs/41803/signedlogcontent/62?urlExpires=2021-03-15T10%3A33%3A53.1605722Z&urlSigningMethod=HMACV1&urlSignature=wdQSJpI%2FMRR0L8CQN8qzg%2Bz%2FD58ic2XdpXQm%2FEYNYVg%3D
https://pipelines.actions.githubusercontent.com/rI6se88BdHGIt3c6pWSIOE7hniT9sK2Yykq02f3adUskIOWLhd/_apis/pipelines/1/runs/41784/signedlogcontent/64?urlExpires=2021-03-15T11%3A51%3A10.5229666Z&urlSigningMethod=HMACV1&urlSignature=PSJmMVUGj9jRbGmfnS0AaHRIx%2FQSfewr3aClt6bhRTA%3D

The common pattern is antrea-agent cannot get up and running after runnning TestFlowAggregator, the case itself may succeed or fail:
```
2021-03-13T06:13:57.0285748Z === RUN   TestFlowAggregator
2021-03-13T06:13:57.0288577Z ##[error]    fixtures.go:133: Creating 'antrea-test' K8s Namespace
2021-03-13T06:13:57.0311588Z ##[error]    fixtures.go:96: Applying Antrea YAML
2021-03-13T06:13:57.9460506Z ##[error]    fixtures.go:100: Waiting for all Antrea DaemonSet Pods
2021-03-13T06:13:58.9598129Z ##[error]    fixtures.go:104: Checking CoreDNS deployment
2021-03-13T06:14:01.9906449Z ##[error]    fixtures.go:160: Applying flow aggregator YAML with ipfix collector address: 172.18.0.4:4739:tcp
2021-03-13T06:14:05.3566800Z ##[error]    fixtures.go:172: Deploying flow exporter with collector address: 10.96.63.61:4739:tcp
2021-03-13T06:15:36.9364958Z ##[error]    flowaggregator_test.go:110: Error when setting up test: error when restarting antrea-agent Pod: antrea-agent DaemonSet not ready within 1m30s
2021-03-13T06:15:36.9368127Z --- FAIL: TestFlowAggregator (99.91s)
2021-03-13T06:15:36.9369146Z === RUN   TestReplaceFieldValue
2021-03-13T06:15:36.9372023Z --- PASS: TestReplaceFieldValue (0.00s)
2021-03-13T06:15:36.9392459Z === RUN   TestIPSecTunnelConnectivity
2021-03-13T06:15:36.9395283Z ##[error]    fixtures.go:36: Skipping test for the 'kind' provider: IPSec tunnel does not work with Kind
2021-03-13T06:15:36.9398119Z --- SKIP: TestIPSecTunnelConnectivity (0.00s)
2021-03-13T06:15:36.9399464Z === RUN   TestIPSecDeleteStaleTunnelPorts
2021-03-13T06:15:36.9401897Z ##[error]    fixtures.go:36: Skipping test for the 'kind' provider: IPSec tunnel does not work with Kind
2021-03-13T06:15:36.9407904Z --- SKIP: TestIPSecDeleteStaleTunnelPorts (0.00s)
2021-03-13T06:15:36.9409278Z === RUN   TestNetworkPolicyStats
2021-03-13T06:15:36.9411329Z ##[error]    fixtures.go:133: Creating 'antrea-test' K8s Namespace
2021-03-13T06:15:36.9413605Z ##[error]    fixtures.go:96: Applying Antrea YAML
2021-03-13T06:17:08.0802565Z ##[error]    networkpolicy_test.go:40: Error when setting up test: error when waiting for antrea-agent rollout to complete
2021-03-13T06:17:08.0859094Z --- FAIL: TestNetworkPolicyStats (91.14s)
2021-03-13T06:17:08.0860619Z === RUN   TestDifferentNamedPorts
2021-03-13T06:17:08.0862750Z ##[error]    fixtures.go:133: Creating 'antrea-test' K8s Namespace
2021-03-13T06:17:08.0865191Z ##[error]    fixtures.go:96: Applying Antrea YAML
2021-03-13T06:18:39.0666049Z ##[error]    networkpolicy_test.go:155: Error when setting up test: error when waiting for antrea-agent rollout to complete
2021-03-13T06:18:39.0686473Z --- FAIL: TestDifferentNamedPorts (90.99s)
```

```
--- PASS: TestFlowAggregator (209.81s)
    --- PASS: TestFlowAggregator/IntraNodeFlows (11.19s)
    --- PASS: TestFlowAggregator/InterNodeFlows (11.24s)
    --- PASS: TestFlowAggregator/LocalServiceAccess (11.10s)
    --- PASS: TestFlowAggregator/RemoteServiceAccess (11.10s)
=== RUN   TestReplaceFieldValue
--- PASS: TestReplaceFieldValue (0.00s)
=== RUN   TestIPSecTunnelConnectivity
    fixtures.go:133: Creating 'antrea-test' K8s Namespace
    fixtures.go:96: Applying Antrea YAML
    fixtures.go:100: Waiting for all Antrea DaemonSet Pods
    ipsec_test.go:66: Error when setting up test: antrea-agent DaemonSet not ready within 1m30s
--- FAIL: TestIPSecTunnelConnectivity (91.18s)
=== RUN   TestIPSecDeleteStaleTunnelPorts
    fixtures.go:133: Creating 'antrea-test' K8s Namespace
    fixtures.go:96: Applying Antrea YAML
    ipsec_test.go:100: Error when setting up test: error when waiting for antrea-agent rollout to complete
--- FAIL: TestIPSecDeleteStaleTunnelPorts (91.16s)
```
**To Reproduce**
Run e2e test

**Expected**
A test case should clean up its side effect, not affecting other cases.

> @srikartati maybe the hack that was used to ensure correct restart order between agent and aggregator?
> please prioritize this so that PRs don't get stuck because of CI failures.

Flow aggregator starting with no sleep affects only its yaml, but in this case, flow aggregator started fine and the error is related to Antrea Agent not restarting again properly during `TestFlowAggregator` or after the test. Need to look at logs to see why Antrea Deamonset did not come up properly.

```
2021-03-13T06:14:01.9906449Z ##[error]    fixtures.go:160: Applying flow aggregator YAML with ipfix collector address: 172.18.0.4:4739:tcp
2021-03-13T06:14:05.3566800Z ##[error]    fixtures.go:172: Deploying flow exporter with collector address: 10.96.63.61:4739:tcp
2021-03-13T06:15:36.9364958Z ##[error]    flowaggregator_test.go:110: Error when setting up test: error when restarting antrea-agent Pod: antrea-agent DaemonSet not ready within 1m30s
2021-03-13T06:15:36.9368127Z --- FAIL: TestFlowAggregator (99.91s)
```

Tried to reproduce this locally to get hold of agent logs when the error has occurred, but I was not successful even after running multiple times. Artifacts in tests do not seem to store the logs when the error has occurred on both kind and Jenkins runs. I was thinking of doing dummy PR and turning on `logsExportOnSuccess` to fetch logs for CI runs. Any better ideas on saving logs?

--- a/pkg/agent/flowexporter/exporter/exporter.go
+++ b/pkg/agent/flowexporter/exporter/exporter.go
@@ -80,21 +80,23 @@ const (
 )
 
 type flowExporter struct {
-	connStore         connections.ConnectionStore
-	flowRecords       *flowrecords.FlowRecords
-	process           ipfix.IPFIXExportingProcess
-	elementsListv4    []*ipfixentities.InfoElementWithValue
-	elementsListv6    []*ipfixentities.InfoElementWithValue
-	ipfixSet          ipfix.IPFIXSet
-	numDataSetsSent   uint64 // used for unit tests.
-	templateIDv4      uint16
-	templateIDv6      uint16
-	registry          ipfix.IPFIXRegistry
-	v4Enabled         bool
-	v6Enabled         bool
-	exporterInput     exporter.ExporterInput
-	activeFlowTimeout time.Duration
-	idleFlowTimeout   time.Duration
+	connStore                 connections.ConnectionStore
+	flowRecords               *flowrecords.FlowRecords
+	process                   ipfix.IPFIXExportingProcess
+	elementsListv4            []*ipfixentities.InfoElementWithValue
+	elementsListv6            []*ipfixentities.InfoElementWithValue
+	ipfixSet                  ipfix.IPFIXSet
+	numDataSetsSent           uint64 // used for unit tests.
+	templateIDv4              uint16
+	templateIDv6              uint16
+	registry                  ipfix.IPFIXRegistry
+	v4Enabled                 bool
+	v6Enabled                 bool
+	exporterInput             exporter.ExporterInput
+	activeFlowTimeout         time.Duration
+	idleFlowTimeout           time.Duration
+	enableTLSToFlowAggregator bool
+	k8sClient                 kubernetes.Interface
 }
 
 func genObservationID() (uint32, error) {
@@ -107,40 +109,13 @@ func genObservationID() (uint32, error) {
 	return h.Sum32(), nil
 }
 
-func prepareExporterInputArgs(collectorAddr, collectorProto string, enableTLSToFlowAggregator bool, k8sClient kubernetes.Interface) (exporter.ExporterInput, error) {
+func prepareExporterInputArgs(collectorAddr, collectorProto string) (exporter.ExporterInput, error) {
 	expInput := exporter.ExporterInput{}
 	var err error
 	// Exporting process requires domain observation ID.
 	expInput.ObservationDomainID, err = genObservationID()
 	if err != nil {
 		return expInput, err
-	}
-
-	if enableTLSToFlowAggregator {
-		// if CA certificate, client certificate and key do not exist during initialization,
-		// it will retry to obtain the credentials in next export cycle
-		expInput.CACert, err = getCACert(k8sClient)
-		if err != nil {
-			return expInput, fmt.Errorf("cannot retrieve CA cert: %v", err)
-		}
-		expInput.ClientCert, expInput.ClientKey, err = getClientCertKey(k8sClient)
-		if err != nil {
-			return expInput, fmt.Errorf("cannot retrieve client cert and key: %v", err)
-		}
-		// TLS transport does not need any tempRefTimeout as it is applicable only
-		// for TCP transport, so sending 0.
-		expInput.TempRefTimeout = 0
-		expInput.IsEncrypted = true
-	} else if collectorProto == "tcp" {
-		// TCP transport does not need any tempRefTimeout, so sending 0.
-		// tempRefTimeout is the template refresh timeout, which specifies how often
-		// the exporting process should send the template again.
-		expInput.TempRefTimeout = 0
-		expInput.IsEncrypted = false
-	} else {
-		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
-		expInput.TempRefTimeout = 1800
-		expInput.IsEncrypted = false
 	}
 	expInput.CollectorAddress = collectorAddr
 	expInput.CollectorProtocol = collectorProto
@@ -157,20 +132,22 @@ func NewFlowExporter(connStore connections.ConnectionStore, records *flowrecords
 	registry.LoadRegistry()
 
 	// Prepare input args for IPFIX exporting process.
-	expInput, err := prepareExporterInputArgs(collectorAddr, collectorProto, enableTLSToFlowAggregator, k8sClient)
+	expInput, err := prepareExporterInputArgs(collectorAddr, collectorProto)
 	if err != nil {
 		return nil, err
 	}
 	return &flowExporter{
-		connStore:         connStore,
-		flowRecords:       records,
-		registry:          registry,
-		v4Enabled:         v4Enabled,
-		v6Enabled:         v6Enabled,
-		exporterInput:     expInput,
-		activeFlowTimeout: activeFlowTimeout,
-		idleFlowTimeout:   idleFlowTimeout,
-		ipfixSet:          ipfix.NewSet(false),
+		connStore:                 connStore,
+		flowRecords:               records,
+		registry:                  registry,
+		v4Enabled:                 v4Enabled,
+		v6Enabled:                 v6Enabled,
+		exporterInput:             expInput,
+		activeFlowTimeout:         activeFlowTimeout,
+		idleFlowTimeout:           idleFlowTimeout,
+		ipfixSet:                  ipfix.NewSet(false),
+		enableTLSToFlowAggregator: enableTLSToFlowAggregator,
+		k8sClient:                 k8sClient,
 	}, nil
 }
 
@@ -212,10 +189,36 @@ func (exp *flowExporter) Export() {
 
 func (exp *flowExporter) initFlowExporter() error {
 	var err error
-	exp.process, err = ipfix.NewIPFIXExportingProcess(exp.exporterInput)
+	if exp.enableTLSToFlowAggregator {
+		// if CA certificate, client certificate and key do not exist during initialization,
+		// it will retry to obtain the credentials in next export cycle
+		exp.exporterInput.CACert, err = getCACert(exp.k8sClient)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve CA cert: %v", err)
+		}
+		exp.exporterInput.ClientCert, exp.exporterInput.ClientKey, err = getClientCertKey(exp.k8sClient)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve client cert and key: %v", err)
+		}
+		// TLS transport does not need any tempRefTimeout, so sending 0.
+		exp.exporterInput.TempRefTimeout = 0
+		exp.exporterInput.IsEncrypted = true
+	} else if exp.exporterInput.CollectorProtocol == "tcp" {
+		// TCP transport does not need any tempRefTimeout, so sending 0.
+		// tempRefTimeout is the template refresh timeout, which specifies how often
+		// the exporting process should send the template again.
+		exp.exporterInput.TempRefTimeout = 0
+		exp.exporterInput.IsEncrypted = false
+	} else {
+		// For UDP transport, hardcoding tempRefTimeout value as 1800s.
+		exp.exporterInput.TempRefTimeout = 1800
+		exp.exporterInput.IsEncrypted = false
+	}
+	expProcess, err := ipfix.NewIPFIXExportingProcess(exp.exporterInput)
 	if err != nil {
 		return fmt.Errorf("error when starting exporter: %v", err)
 	}
+	exp.process = expProcess
 	if exp.v4Enabled {
 		templateID := exp.process.NewTemplateID()
 		exp.templateIDv4 = templateID

--- a/test/e2e/fixtures.go
+++ b/test/e2e/fixtures.go
@@ -143,42 +143,43 @@ func setupTest(tb testing.TB) (*TestData, error) {
 func setupTestWithIPFIXCollector(tb testing.TB) (*TestData, error, bool) {
 	// TODO: remove hardcoding to IPv4 after flow aggregator supports IPv6
 	isIPv6 := false
-	if _, err := setupTest(tb); err != nil {
+	data, err := setupTest(tb)
+	if err != nil {
 		return nil, err, isIPv6
 	}
 	// Create pod using ipfix collector image
-	if err := testData.createPodOnNode("ipfix-collector", "", ipfixCollectorImage, nil, nil, nil, nil, true, nil); err != nil {
+	if err = data.createPodOnNode("ipfix-collector", "", ipfixCollectorImage, nil, nil, nil, nil, true, nil); err != nil {
 		tb.Errorf("Error when creating the ipfix collector Pod: %v", err)
 	}
-	ipfixCollectorIP, err := testData.podWaitForIPs(defaultTimeout, "ipfix-collector", testNamespace)
+	ipfixCollectorIP, err := data.podWaitForIPs(defaultTimeout, "ipfix-collector", testNamespace)
 	if err != nil || len(ipfixCollectorIP.ipStrings) == 0 {
 		tb.Errorf("Error when waiting to get ipfix collector Pod IP: %v", err)
 		return nil, err, isIPv6
 	}
 	ipStr := ipfixCollectorIP.ipv4.String()
 	ipfixCollectorAddr := fmt.Sprintf("%s:%s:tcp", ipStr, ipfixCollectorPort)
+
+	faClusterIPAddr := ""
 	tb.Logf("Applying flow aggregator YAML with ipfix collector address: %s", ipfixCollectorAddr)
-	faClusterIP, err := testData.deployFlowAggregator(ipfixCollectorAddr)
+	faClusterIP, err := data.deployFlowAggregator(ipfixCollectorAddr)
 	if err != nil {
 		return testData, err, isIPv6
 	}
-
-	faClusterIPAddr := ""
 	if testOptions.providerName == "kind" {
 		// In Kind cluster, there are issues with DNS name resolution on worker nodes.
 		// Please note that CoreDNS services are forced on to control-plane Node.
 		faClusterIPAddr = fmt.Sprintf("%s:%s:tcp", faClusterIP, ipfixCollectorPort)
 	}
 	tb.Logf("Deploying flow exporter with collector address: %s", faClusterIPAddr)
-	if err = testData.deployAntreaFlowExporter(faClusterIPAddr); err != nil {
-		return testData, err, isIPv6
+	if err = data.deployAntreaFlowExporter(faClusterIPAddr); err != nil {
+		return data, err, isIPv6
 	}
 
 	tb.Logf("Checking CoreDNS deployment")
-	if err = testData.checkCoreDNSPods(defaultTimeout); err != nil {
-		return testData, err, isIPv6
+	if err = data.checkCoreDNSPods(defaultTimeout); err != nil {
+		return data, err, isIPv6
 	}
-	return testData, nil, isIPv6
+	return data, nil, isIPv6
 }
 
 func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs bool) {
@@ -304,7 +305,7 @@ func exportLogs(tb testing.TB, data *TestData, logsSubDir string, writeNodeLogs 
 }
 
 func teardownFlowAggregator(tb testing.TB, data *TestData) {
-	if err := testData.gracefulExitFlowAggregator(testOptions.coverageDir); err != nil {
+	if err := data.gracefulExitFlowAggregator(testOptions.coverageDir); err != nil {
 		tb.Fatalf("Error when gracefully exiting Flow Aggregator: %v", err)
 	}
 	tb.Logf("Deleting '%s' K8s Namespace", flowAggregatorNamespace)

--- a/test/e2e/flowaggregator_test.go
+++ b/test/e2e/flowaggregator_test.go
@@ -109,8 +109,8 @@ func TestFlowAggregator(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Error when setting up test: %v", err)
 	}
-	defer teardownFlowAggregator(t, data)
 	defer teardownTest(t, data)
+	defer teardownFlowAggregator(t, data)
 
 	podAIP, podBIP, podCIP, svcB, svcC, err := createPerftestPods(data)
 	if err != nil {

--- a/test/e2e/framework.go
+++ b/test/e2e/framework.go
@@ -681,7 +681,8 @@ func (data *TestData) waitForAntreaDaemonSetPods(timeout time.Duration) error {
 		return true, nil
 	})
 	if err == wait.ErrWaitTimeout {
-		return fmt.Errorf("antrea-agent DaemonSet not ready within %v", defaultTimeout)
+		_, stdout, _, _ := provider.RunCommandOnNode(controlPlaneNodeName(), fmt.Sprintf("kubectl -n %s describe pod", antreaNamespace))
+		return fmt.Errorf("antrea-agent DaemonSet not ready within %v; kubectl describe pod output: %v", defaultTimeout, stdout)
 	} else if err != nil {
 		return err
 	}

